src/main/java/core/basesyntax/service/ShopServiceImpl.java

```
@@ -0,0 +1,28 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ShopServiceImpl implements ShopService {
+    private final Map<String, Integer> stock = new HashMap<>();
+    private final OperationStrategy strategy;
+
+    public ShopServiceImpl(OperationStrategy strategy) {
+        this.strategy = strategy;
+    }
+
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+        for (FruitTransaction transaction : transactions) {
+            strategy.getHandler(transaction.getOperation()).handle(transaction, stock);
```

Inside process(...) you call `strategy.getHandler(transaction.getOperation()).handle(transaction, stock);` without checking the returned handler. OperationStrategyImpl currently does a raw `handlers.get(operation)` and can return null; the strategy should either throw an informative error or you should check for null and throw an informative RuntimeException before calling handle. Example: retrieve the handler into a variable, if (handler == null) throw new RuntimeException("No handler for operation: " + operation); See OperationStrategyImpl behavior.